File: backend/src/routes/notifications.js
const express = require("express");
const router = express.Router();
const pool = require("../db");

// GET /api/notifications
router.get("/", async (req, res) => {
  try {
    const [upcomingRes, overdueRes] = await Promise.all([
      pool.query(
        `SELECT fr.*, rs.model_name, c.name AS customer_name, c.phone AS customer_phone
         FROM filter_replacements fr
         JOIN ro_systems rs ON fr.ro_system_id = rs.id
         JOIN customers c ON rs.customer_id = c.id
         WHERE fr.next_due_date BETWEEN CURRENT_DATE AND CURRENT_DATE + INTERVAL '7 days'
         ORDER BY fr.next_due_date ASC`
      ),
      pool.query(
        `SELECT fr.*, rs.model_name, c.name AS customer_name, c.phone AS customer_phone
         FROM filter_replacements fr
         JOIN ro_systems rs ON fr.ro_system_id = rs.id
         JOIN customers c ON rs.customer_id = c.id
         WHERE fr.next_due_date < CURRENT_DATE
         ORDER BY fr.next_due_date ASC`
      ),
    ]);

    res.json({
      upcoming: upcomingRes.rows,
      overdue: overdueRes.rows,
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: "Failed to fetch notifications" });
  }
});

module.exports = router;

File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func, text

from ..database import get_db
from ..models import FilterReplacements, ROSystems, Customers
from ..schemas import NotificationsOut, NotificationItem
from ..auth import get_current_user

router = APIRouter(prefix="/api/notifications", tags=["notifications"])


@router.get("/", response_model=NotificationsOut)
def get_notifications(
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user),
):
    # Upcoming — due within next 7 days
    upcoming_rows = (
        db.query(
            FilterReplacements,
            ROSystems.model_name,
            Customers.name.label("customer_name"),
            Customers.phone.label("customer_phone"),
        )
        .join(ROSystems, FilterReplacements.ro_system_id == ROSystems.id)
        .join(Customers, ROSystems.customer_id == Customers.id)
        .filter(
            FilterReplacements.next_due_date.between(
                func.current_date(),
                func.current_date() + text("INTERVAL '7 days'"),
            )
        )
        .order_by(FilterReplacements.next_due_date.asc())
        .all()
    )

    # Overdue
    overdue_rows = (
        db.query(
            FilterReplacements,
            ROSystems.model_name,
            Customers.name.label("customer_name"),
            Customers.phone.label("customer_phone"),
        )
        .join(ROSystems, FilterReplacements.ro_system_id == ROSystems.id)
        .join(Customers, ROSystems.customer_id == Customers.id)
        .filter(FilterReplacements.next_due_date < func.current_date())
        .order_by(FilterReplacements.next_due_date.asc())
        .all()
    )

    def make_items(rows):
        items = []
        for fr, model_name, customer_name, customer_phone in rows:
            items.append(
                NotificationItem(
                    id=fr.id,
                    ro_system_id=fr.ro_system_id,
                    filter_type=fr.filter_type,
                    replaced_date=fr.replaced_date,
                    next_due_date=fr.next_due_date,
                    cost=float(fr.cost) if fr.cost else None,
                    notes=fr.notes,
                    created_at=fr.created_at,
                    model_name=model_name,
                    customer_name=customer_name,
                    customer_phone=customer_phone,
                )
            )
        return items

    return NotificationsOut(
        upcoming=make_items(upcoming_rows),
        overdue=make_items(overdue_rows),
    )
